### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDeps** [bcgsc/NanoSim#172]
Hi, In **NanoSim**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using

```
_anaconda_depends=2019.03=py37_0
_ipyw_jlab_nb_ext_conf=0.1.0=py37_0
_libgcc_mutex=0.1=main
genometools-genometools
htseq=0.11.3=py37hb3f55d8_0
joblib=1.1.0=pyhd3eb1b0_0
last=874=h8b12597_3
minimap2=2.17=h8b12597_1
numpy=1.21.5=pypi_0
numpy-base=1.21.2=py37h79a1101_0
pybedtools=0.8.1=py37he513fc3_0
pysam=0.15.3=py37hda2845c_1
samtools
scikit-learn=0.22.1=pypi_0
scipy=1.7.3=py37hc147768_0
six=1.16.0=pyhd3eb1b0_1
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **joblib** can be changed to *>=0.3.2d.dev,<=0.7.0d*.
The version constraint of dependency **joblib** can be changed to *>=0.3a.dev,<=1.1.0*.
The version constraint of dependency **scikit-learn** can be changed to *>=0.15.0,<=0.20.4*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details><summary>The calling methods from the joblib</summary>
 <pre>joblib.load
joblib.dump
</pre>
</details>
<details><summary>The calling methods from the scikit-learn</summary>
 <pre>sklearn.neighbors.KernelDensity.fit
joblib.dump
random.choice
joblib.load
sklearn.neighbors.KernelDensity
</pre>
</details>
<details><summary>The calling methods from the all methods</summary>
 <pre>poisgeom_gen.cdf
hp_length_hist.keys
list_coords.append
sys.stdout.close
seq_dict.keys
mis_pos.append
iv.features.steps
metagenome_list.values
numpy.random.poisson
mixed_model.trunc_lognorm_rvs
header.strip.split
hp_arr.append
urllib.request.Request.add_header
add_abundance_var.items
numpy.floor
r.transcript_compatibility.append
match.start
seg_error_count_list.append
extract_read
argparse.ArgumentParser.parse_args
seg_par
x.get
read_count_unique.items.items
EM_meta.items
enumerate
file_handler.gzopen.write
e_dict.keys
reversed
supp_aln.split
add_abundance_var.keys
pos.append
weigeom_gen
kde2d
ref.split
reverse_complement
mixed_model.pois_geom
line.strip.split
assign_species
qend.qstart.aln.query_name.quant_dic.append
err_quals.append
line.split.strip
pysam.Fastafile
multiprocessing.Pool.close
os.path.realpath
m_model.keys
textwrap.dedent
IR_markov.readline
round
kde.sample
file_handler.gzopen.close
seqs.append
tname.split.startswith
weigeom_gen.cdf
g_list.items
g_alnm.lower.endswith
abun_var.append
file_path.lower.endswith
trunc_lognorm_gen.rvs
seg_idx.ref_length_list.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
subparsers.add_parser.add_argument
next.split
gap_base_qual_list.append
profile.readline
inflate_abun
parse_cs
fs2.intersection
int
feature.name.split
seq.upper
random.random.start
k1.match_list.values
get_length_kde
last.partition
numpy.random.geometric
s.seq_len.values
set.add
scipy.stats.expon.rvs
gap_list.append
ir_list.append
line.str.strip.split
joblib.load
int.keys
primary_trx.dict_ir_info.append
random.random
model_intron_retention.intron_retention
e_count.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
procs.append
print
next.strip
i.split
dna_nd_par
feature_id.dict_intron_info.append
HTSeq.GFF_Reader
diff.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
numpy.cumsum
gap_mutated.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist.extend
t_alnm.lower.endswith
seg.append
is_equivalent_hit
match.group
EM_meta.values
multiprocessing.Pool.join
gap_mutated.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist.pop
time.strftime
cigar_parser
multi_mapping_list.items
case_convert
parse_paf
line.split.split
fq_path.replace
abs
read_count_unique.items.values
aligned_subfiles.append
list_intervals.append
sys.stdout.write
multiprocessing.Pool
random.choice
ref_len_aligned.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
error_subfiles.append
len
i.ecdf_key.ecdf_dict.keys
fields.strip.split
open
u_profile.readline
get_polya_len
numpy.random.random
list_states.append
sorted
list_to_range
abun_with_var.items
count_each_bin.values
g_alnm.lower
get_head_tail
ref_len_from_structure
next
sys.stderr.write
e_count.basecaller.read_type.mm.trunc_lognorm_rvs.tolist.pop
open.write
read_list.items
output_list.append
first_error.values
parser.add_subparsers.add_parser
each_ref.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
readfq
tname.split.split
dict.fromkeys
model_fitting.model_fitting
x.get.sort
dna_type_list.readlines
k1.m_model.items
format
HTSeq.GenomicInterval
re.findall
str.maketrans
abun_file.readline
abun_file.readlines
chimeric_info.readline
fq_path.startswith
re.escape
l.extend
multiprocessing.Value
i.seg.split
numpy.exp
numpy.power
cdna_nd_par
fq_path.strip.split
list_op.append
weigeom2_gen
k1.m_ht_list.items
scipy.stats.geom.cdf
mutate_read
sys.exit
urllib.request.Request
pysam.AlignmentFile
align_genome
unaligned_subfiles.append
error_log.write
match.end
u_profile.readline.strip
error_markov.readline
set
not_overlap
get_cs
range
file_handler.gzopen
subprocess.call
new_line.strip.split
line.split.partition
numpy.array.append
EM_meta.keys
s.seq_len.keys
argparse.ArgumentParser.add_argument
HTSeq.GenomicArrayOfSets
alnm.get_tag
get_primary_sam.primary_and_unaligned
list_IR_positions.append
multi_dict_abun.keys
numpy.random.normal
is_full_length
pysam.AlignmentFile.fetch
extract_read_trx
k_of_match_list.match_list.values
make_cdf
random.choices
chimeric_species_count.items
re.split.split
open.close
line.rstrip
math.exp
read.compatibility.append
concatenate_genomes
expected_abun.items
joblib.dump
numpy.log10
f.next.strip.split
compatible_list.append
mixed_model.wei_geom
read_mutated.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
HTSeq.GenomicInterval.copy
file_path.lower
norm_distr.get_nd_par
self.cdf
numpy.ones
math.sqrt
scipy.stats.lognorm.ppf
float
hp_pos.append
gzip.GzipFile
m_ht_list.keys
numpy.random.lognormal
query.strip.split
read_mutated.traslate.traslate
hist.items
head_align_tail_dist.edge_checker
get_besthit_maf.besthit_and_unaligned
getopt.getopt
get_primary_sam.primary_and_unaligned_chimeric
strand_profile.readline
u_profile.readline.strip.split
subparsers.add_parser.print_help
statistics.median
numpy.histogram
multiprocessing.Pool.apply_async
numpy.log
read_ecdf
seg.sort
random.seed
ValueError
aln.reference_name.split
r.get_blocks
list_value.append
besthit_to_histogram.hist
gap_mutated.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
check_print_progress
infile.read
beta_list.append
random.randint
hp_samples.keys
scipy.stats.lognorm.cdf
get_compatibility
seg_length_list.append
quals.append
len.values
match_list.keys
iv.chrom.strip
current_species_base_dict.values
pysam.AlignmentFile.close
unaligned_error_list
strand_profile.readline.split
match_bin.keys
numpy.abs
trunc_lognorm_gen
annot_file_extension.upper
simulation
pysam.Fastafile.fetch
numpy.power.flatten
base_count_unique.items.items
len.split
dict.fromkeys.items
read_mutated.traslate.translate
species.seq_len.values
pysam.AlignmentFile.write
re.split
main
os.path.splitext
base_quota.items
edge_checker
EM_trans.items
species_chrom.split
tail.head.basecaller.read_type.mm.trunc_lognorm_rvs.tolist.pop
prev_error.trans_p.keys
str
t_alnm.lower
error_rate.keys
multiprocessing.Value.get_lock
alnm.infer_read_length
seq_len.keys
line.rstrip.split
comp.get
trx_list.readlines
urllib.request.urlopen
out_aligned_reads.write
head_align_tail_dist.head_align_tail
random.uniform
feature.iv.chrom.strip
list.append
exp_file.readline.strip
list_hist.append
fq_path.strip
add_dict
ecdf_dict.keys
seq_len.values
quals.reverse
norm_distr.get_hpmis_rate
read_profile
math.ceil
numpy.array
six.moves.xrange
os.path.dirname
update_compatibility
sklearn.neighbors.KernelDensity
bw.KernelDensity.fit
sys.stdout.flush
polya_len.tail.head.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
read_histogram
collections.defaultdict
new_line.strip.split.upper
numpy.vstack
random.choice.split
gap_mutated.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist.reverse
ref_len_total.fc.np.abs.argmin
add_match
numpy.random.choice
list
f.write
mod_profile.readline
select_nearest_kde2d
numpy.random.seed
genome_list.readlines
add_abundance_var
simulation_gap
chimeric_info.readline.split
add_intron
mutate_homo
chr
max
line.strip
total_ref_length.append
min
conv_op_to_word
argparse.ArgumentParser.add_subparsers
dic_first_match.values
argparse.ArgumentParser.print_help
feature_id.dict_ref_structure.append
cigar.split
dict
scipy.stats.poisson.pmf
list.remove
aln.get_tag.split
aligned_ref_length.append
seg_error_dict_list.append
out_unaligned_reads.write
sum
random.random.join
error_list
os.remove
EM_trans
mis_pos.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
aln.get_tag
line.str.strip
scipy.optimize.minimize
copy.deepcopy
update_structure
zip
fq_path.replace.strip
new_e_dict.keys
gap_mutated.len.basecaller.read_type.mm.trunc_lognorm_rvs.tolist.append
numpy.arange
drna_nd_par
abun_with_var.values
poisgeom_gen
sklearn.neighbors.KernelDensity.fit
EM_meta
calculate_abundance
join
exp_file.readline
numpy.random.weibull
align_transcriptome
line.str.strip.strip
re.finditer
length.hp_samples.keys
f.next.strip
f.readline
extract_read_pos
itertools.chain.from_iterable
add_abundance_var.values
string.maketrans
join.split
scipy.stats.geom.pmf
multiprocessing.Process
cs.extend
gzip.open
tail.head.basecaller.read_type.mm.trunc_lognorm_rvs.tolist
cigar.split.split
argparse.ArgumentParser
feature_id.split.split
</pre>
</details>

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
